Apps/Database/src/Delegates/DBCommunicationEmailDelegate.cs

```diff
@@ -0,0 +1,112 @@
+        /// Initializes a new instance of the <see cref="DBCommunicationEmailDelegate"/> class.
+        /// </summary>
+        /// <param name="logger">Injected Logger Provider.</param>
+        /// <param name="dbContext">The context to be used when accessing the database.</param>
+        public DBCommunicationEmailDelegate(
+            ILogger<DBNoteDelegate> logger,
+            GatewayDbContext dbContext)
+        {
+            this.logger = logger;
+            this.dbContext = dbContext;
+        }
+
+        /// <inheritdoc />
+        public DBResult<CommunicationEmail> Add(CommunicationEmail communicationEmail, bool commit = true)
+        {
+            this.logger.LogTrace($"Adding Communication Email to DB...");
+            DBResult<CommunicationEmail> result = new DBResult<CommunicationEmail>()
+            {
+                Payload = communicationEmail,
+                Status = DBStatusCode.Deferred,
+            };
+            this.dbContext.CommunicationEmail.Add(communicationEmail);
+            if (commit)
+            {
+                try
+                {
+                    this.dbContext.SaveChanges();
+                    result.Status = DBStatusCode.Created;
+                }
+                catch (DbUpdateException e)
+                {
+                    this.logger.LogError($"Unable to save Communication Email to DB {e}");
+                    result.Status = DBStatusCode.Error;
+                    result.Message = e.Message;
+                }
+            }
+
+            this.logger.LogDebug($"Finished adding Communication in DB");
+            return result;
+        }
+
+        /// <inheritdoc />
+        public List<UserProfile> GetActiveUserProfilesWithoutCommEmailByCommunicationIdAndByCreatedOnOrAfter(Guid communicationId, DateTime? createdOnOrAfter = null, int maxRows = 500)
+        {
+            this.logger.LogTrace($"Getting Communication Emails by Communication Id from DB...");
+
+            // Retrieves the communication emails and user profiles by communicationId and createdOnOrAfter.
+            var commEmails = this.dbContext.CommunicationEmail.Where(c => c.CommunicationId == communicationId).OrderByDescending(c => c.CreatedDateTime).Take(maxRows)
+                    .Include(commEmail => commEmail.UserProfile).Where(c => !createdOnOrAfter.HasValue || c.UserProfile.CreatedDateTime >= createdOnOrAfter).ToList();
+            string hdidsWithMostRecentCreatedCommEmails = string.Empty;
+            if (commEmails.Count > 0)
+            {
+                createdOnOrAfter = commEmails[0].UserProfile.CreatedDateTime;
+                StringBuilder hdids = new StringBuilder();
+                foreach (var commEmail in commEmails)
+                {
+                    hdids.Append(commEmail.UserProfileHdId + "|");
+                }
+
+                hdidsWithMostRecentCreatedCommEmails = hdids.ToString();
+            }
+
+            var userProfiles = this.dbContext.UserProfile
```

Comment:
I think we could likely optimize this query. Can you create a task in azure for the future?

Reply:
Created 8717 for this improvement.